Added routes for signin, signout, signup and currentuser routes

litcodeur opened the pull request:
Route informations
- POST /api/auth/signup to signup users takes parameters of { email, firstName, lastName, password} and ( Editable: creates a JWT for user )
- POST /api/auth/signin to signin users takes parameters of {email,password} and creates a JWT
- GET /api/auth/currentuser gets the information stored in jwt after authorizing if the user is logged in or not
- POST /api/auth/signout removes the session

A couple notes on refactors

1) All the database stuff is now in the database directory so if someone wants to perform database operations just call db = require('/database/models/index') then db.<ModelName>.findOne({}) or db.<ModelName>.create({})

2) All the source code for the server now resides inside of src directory

3) I have used classes to streamline errorhandling, so any requests would have a common error response object. The return object has the following structure
{
   errors: [
                 { message, field? }...
               ]
}

4) If someone wants to create a new path and have custom errors just create a new class that extends the CustomError class inside of the errors directory and then throw the error when a condition applies.

5) Add the currentUser middleware and then the requireAuth middleware on your route to make sure only logged in users are allowed to access the routes
EG: router.post('/somesecretroute',currentUser,requireAuth,(req,res)=>{res.send({very secret information})});

6) All the sequelize config is in the .sequelizerc file and then in the server directory is where all the refrences for sequelize are saved. You might need to run "npx sequlize init" inside the server directory.

Try doing "npx sequelize db:migrate" if it shows no changes to be made in the console. try out "npx sequelize db:migrate:undo:all" to revert the changes and then try migrating again

Any more suggesstions and reviews welcome!

Shah62 committed:
Added routes for signin, signout, signup and currentuser routes

Route informations
- POST /api/auth/signup to signup users takes parameters of { email, firstName, lastName, password} and ( Editable: creates a JWT for user )
- POST /api/auth/signin to signin users takes parameters of {email,password} and creates a JWT
- GET /api/auth/currentuser gets the information stored in jwt after authorizing if the user is logged in or not
- POST /api/auth/signout removes the session

A couple notes on refactors

1) All the database stuff is now in the database directory so if someone wants to perform database operations just call db = require('/database/models/index') then db.<ModelName>.findOne({}) or db.<ModelName>.create({})

2) All the source code for the server now resides inside of src directory

3) I have used classes to streamline errorhandling, so any requests would have a common error response object. The return object has the following structure
{
   errors: [
                 { message, field? }...
               ]
}

4) If someone wants to create a new path and have custom errors throw just extend the CustomError class inside of the errors directory and then throw the error when a condition applies.

5) Add the currentUser middleware and then the requireAuth middleware on your route to make sure only logged in users are allowed to access the routes
EG: router.post('/somesecretroute',currentUser,requireAuth,(req,res)=>{res.send({very secret information})});

6) All the sequelize config is in the .sequelizerc file and then in the server directory is where all the refrences for sequelize are saved

Try doing "npx sequelize db:migrate" if it shows no changes to be made in the console. try out "npx sequelize db:migrate:undo:all" to revert the changes and then try migrating again

Any more suggesstions and reviews welcome!

skassam21 commented on .gitignore:
```diff
@@ -18,6 +18,6 @@ npm-debug.log*
-
+server/.env
+server/.env
+server/.env
```
We should only need one of these lines :)

skassam21 commented on server/database/config/config.js:
```diff
@@ -1,18 +1,18 @@
     database: process.env.DATABASE,
-    host: "127.0.0.1",
-    dialect: "postgres",
+    host: '127.0.0.1',
```
@MaharjanR I think @shah62 is using a different linter / beautifier than you - we should get on the same page so that both of you have the same formatted code

skassam21 commented on server/database/models/user.js:
```diff
@@ -0,0 +1,102 @@
+    delete values.lastName;
+    delete values.firstName;
+    return values;
+  }
```
Nice job on these class methods 👍

skassam21 commented on server/src/app.js:
```diff
@@ -3,9 +3,16 @@ const express = require("express");
 const logger = require("morgan");
+require("express-async-errors");
+const errorHandler = require("./middleware/error-handler");
+const cookiesession = require("cookie-session");
```
should this be camelCase, so cookieSession?

skassam21 commented on server/src/middleware/error-handler.js:
```diff
@@ -0,0 +1,39 @@
+const errorHandler = (err, req, res, next) => {
+  
+  // Check if the error extends the CustomError Class which would make it an instance of it
+  if (err instanceof CustomError) {
```
Interesting approach, actually haven't seen people use this object oriented approach to throwing errors in middlewares - I like it

skassam21 commented on server/src/routes/current-user.js:
```diff
@@ -0,0 +1,24 @@
+router.get(
+  "/api/auth/currentuser",
+  currentUser, //A middleware to check for currentUser stored inside of jwt value of cookie and assign it to a property of currentUser on the request object
+  requireAuth, //A middleware to confirm that a user is logged in otherwise would throw a Not authorized error
```
I feel like we will always use currentUser and requireAuth together right? I can't see a use case when we don't. Should we combine those middlewares to make things a bit more cleaner?

skassam21 commented on server/src/routes/ping.js:
```diff
@@ -0,0 +1,20 @@
+
+const router = express.Router();
+
+router.post('/', (req, res) => {
```
We don't really need these routes anymore, so you can remove them

skassam21 commented on server/src/routes/signin.js:
```diff
@@ -0,0 +1,78 @@
+require("dotenv").config();
+
+router.post(
+  "/api/auth/signin",
```
I think for our routes, we don't need to use /api/auth as a prefix, maybe just /api/

skassam21 commented on server/src/routes/signin.js:
```diff
@@ -0,0 +1,78 @@
+    // Set a cookie called jwt inside the session
+    req.session = {
+      jwt: userJWT,
+    };
```
Did you confirm if you use express sessions it can handle multiple users to sign up at the same time?

litcodeur replied:
I forgot my project was based off of cookie-session and not express-session. I tried using two instances of a rest client and it works great on both.

skassam21 reviewed (commented):
Overall looks great! In terms of overall structure, I'm just a bit worried about the routes folder, can we try using this structure for routes and controllers:
https://www.freecodecamp.org/news/how-to-write-a-production-ready-node-and-express-app-f214f0b17d8c/

Shah62 committed:
Implemented a MVC type architecture, authentication is now handled by one middleware and added an errror for when DB connection fails

- requireAuth is the only middleware required to check if the user is authorized and is logged in or not.
- Removed the ping and index paths
- All new routes from /api/auth/ redirected to /api/
- Fixed the gitignore file
